### PR TITLE
feature: [spv-front] Support for OpAtomicCompareExchange, fixes #6296

### LINKED ISSUE #6296: [naga spv-in] Support for OpAtomicCompareExchange
**Connections**

Part of
* https://github.com/gfx-rs/wgpu/issues/4489

PR

* https://github.com/gfx-rs/wgpu/pull/6590 

**Description**

[`OpAtomicCompareExchange`](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#OpAtomicCompareExchange) was removed from initial SPIR-V atomics support because it requires more infrastructure to support type-upgrading. Previous (now reverted) commits exist as a starting point to finish this work.

### PR DESCRIPTION
**Pull request by schell**
**Connections**
* #6296 
* #4489 

**Description**
Adds support for parsing and executing OpAtomicCompareExchange in the SPIR-V frontend. 
This concludes the work to support atomics in the SPIR-V frontend, excluding test clean-up.

**Testing**

```
RUST_LOG=trace cargo watch -x 'test -p naga --features spv-in,wgsl-in,wgsl-out -- --nocapture atomic_compare_exchange'
```

**Checklist**

- [ ] First, merge #6682.
- [x] Run `cargo fmt`.
- [ ] Run `taplo format`.
- [x] Run `cargo clippy`. If applicable, add:
  - [ ] `--target wasm32-unknown-unknown`
  - [ ] `--target wasm32-unknown-emscripten`
- [x] Run `cargo xtask test` to run tests.
- [x] Add change to `CHANGELOG.md`. See simple instructions inside file.

### CODE REVIEW

File: naga/src/front/spv/mod.rs
Comment (jimblandy) [inline naga/src/front/spv/mod.rs]: The SPIR-V spec calls this "Scope `<id>`: Memory". So maybe this should be called `_memory_scope_id`.

**Review by jimblandy — COMMENTED** [review]
Okay, I got into reviewing this, but ended up filing other stuff I noticed in the process (#6670 and #6671). I'll pick it up again tomorrow.

**Review by jimblandy — CHANGES_REQUESTED** [review]
The changes to `type_gen.rs` are not the way we should handle this. That file should remain unchanged by this PR.

Instead, `BlockContext` should borrow a shared reference to `Module::special_types` in addition to all the other bits and pieces it borrows, and then it should simply look up the appropriate `crate::PredeclaredType::AtomicCompareExchangeWeakResult` key in `SpecialTypes::predeclared_types`. It knows it will find the entry, because `parse` put them both there.

**Comment by jimblandy** [comment]
@schell Just a heads up, I pushed a tiny cleanup commit to this branch.

File: naga/src/front/spv/mod.rs
Comment (jimblandy) [inline naga/src/front/spv/mod.rs]: Like the other insertions into `upgrade_atomics`, this should use `insert`, not `extend`, and apply the `?` operator to the return value of `get_contained_global_variable`, so that if `p_exp_h` cannot be traced back to some global variable, we get an error, rather than simply failing to add the variable to `upgrade_atomics`.

**Review by jimblandy — CHANGES_REQUESTED** [review]

**Comment by jimblandy** [comment]
The way this PR causes the atomic result types to be included in all modules built from SPIR-V source is not great. But it's not straightforward to address unless we fix #6679.

**Comment by jimblandy** [comment]
Okay, I've put up #6682 to fix #6679, rebased this PR on top of that, and then made the atomic struct creation work the way it should. This PR now makes no changes to the snapshot output files.

**Comment by jimblandy** [comment]
This PR cannot be merged, in its current form, until #6682 is merged.

**Review by jimblandy — DISMISSED** [review]
If I understand correctly, this is the final phase of your work on SPIR-V atomic input. If that's the case, then it's time to remove the tests in `front/spv/mod.rs`, and turn `atomic_compare_exchange.spv` into an ordinary snapshot test.

**Comment by schell** [comment]
Cleaning up the tests is still a part of my intended work. My grant project goes to the end of the year (and I've re-applied for 2025 as well).

File: naga/src/front/spv/mod.rs
Comment (jimblandy) [inline naga/src/front/spv/mod.rs]: (carried over) The SPIR-V spec calls this "Scope <id>: Memory". So maybe this should be called `_memory_scope_id`.

File: naga/src/front/spv/mod.rs
Comment (jimblandy) [inline naga/src/front/spv/mod.rs]: (carried forward) Like the other insertions into `upgrade_atomics`, this should use `insert`, not `extend`, and apply the `?` operator to the return value of `get_contained_global_variable`, so that if `p_exp_h` cannot be traced back to some global variable, we get an error, rather than simply failing to add the variable to `upgrade_atomics`.

**Review by jimblandy — CHANGES_REQUESTED** [review]
It's okay to address the tests in a subsequent PR. For now, here are the only remaining items to be addressed. Pretty minor, this looks good otherwise.

**Comment by schell** [comment]
Your last two comments have been addressed @jimblandy, I think we're g2g 👍 .

**Review by jimblandy — APPROVED** [review]
Looks great!

**Comment by jimblandy** [comment]
Note, this is still blocked on #6682.

**Comment by schell** [comment]
Thank you @jimblandy ! 🙇

**Pull request merged**
